### DISCUSSION TITLE
GitHub CI Matrix order?

### DISCUSSION THREAD

**Discussion by pepoluan** [aio-libs/aiosmtpd#331] [Ideas]
I wonder if swapping these lines:

```yaml
        os: [ "macos-11", "macos-12", "ubuntu-18.04", "ubuntu-20.04", "ubuntu-22.04", "windows-latest" ]
        python-version: [ "3.7", "3.8", "3.9", "3.10", "3.11", "pypy3.6", "pypy3.7", "pypy3.8" ]
```

to become this:

```yaml
        python-version: [ "3.7", "3.8", "3.9", "3.10", "3.11", "pypy3.6", "pypy3.7", "pypy3.8" ]
        os: [ "macos-11", "macos-12", "ubuntu-18.04", "ubuntu-20.04", "ubuntu-22.04", "windows-latest" ]
```

will make GitHub CI a wee bit faster?

I don't know how GitHub CI builds its matrix ... so my guess with the second format if will make a set of jobs, all Python 3.7 for various OSes, then all 3.8, etc. Thus is any OS is particularly slow (\*cough\*mac\*cough\*) that job can trundle by while faster jobs can overtake it...

Unless GitHub is smart enough to detect "my my this os is a bit slow I'm gonna jump to the next os" ... or just do the jobs randomly ...

I dunno.

**Comment by waynew** [comment]
I *believe* that it builds out the entire matrix and launches jobs as runners become available. Because we're a FLOSS project we get free minutes but they're at a lower priority IIRC.

**Reply by pepoluan** [reply]
Ahahah yeah it's not deterministic.

I shall not spend this effort then.
